crates/sui-indexer/src/handlers/pruner.rs

```diff
@@ -108,6 +110,15 @@ impl Pruner {
     }
 
     pub async fn start(&self, cancel: CancellationToken) -> IndexerResult<()> {
+        let store_clone = self.store.clone();
+        let retention_policies = self.retention_policies.policies.clone();
+        let cancel_clone = cancel.clone();
+        spawn_monitored_task!(update_watermarks_lower_bounds_task(
```

Comment:
any reason of having a separate task for watermark updating instead of doing it together with pruning, which seems enough and more efficient ?

Reply:
i might've overindexed on devnet, which can have multiple epochs happening within a few minutes of each other. in prod this would only occur once a day. but i'm not sure how it would be more efficient to have it all done in the same task though. in the sense that while we're pruning, we need to update watermarks again, we'll have to wait for the pruner to complete its current pruning task before updating to new watermarks

Reply:
iiuc the main difference is, if we want to update partial pruning results to watermarks at higher frequency -- I think it's okay skip partial updates, and only updates for example when the pruner finishes a "pruning task" for example, a checkpoint.

Reply:
I think that makes sense. In fact, we may not need to report partial updates at all, as the existing data would serve as the true lower bound for `pruner`. I would argue that it's still preferable to have these in separate tasks though. We have two actions:

1. on epoch change, set `reader_lo` and `timestamp` for every watermark table
2. prune if earliest existing data in table is less than `reader_lo`

When we bundle them together, in devnet and in our tests, or any situation where epoch changes can happen before the pruner completes its pruning task, what we'll see is that the available data range increases as pruner lags behind, because it's still in the middle of pruning even though it needs to update `reader_lo` and `timestamp`. Even though this is unlikely to happen in prod, I would prefer if the pruner could robustly update the `reader_lo` watermark independently from the actual pruning work.